### PR TITLE
Add stateRoot when generating a block - Closes #6752

### LINKED ISSUE #6752: Add stateRoot when generating a block
### Description

State root must be calculated and add it to the block header.

- Define sparse merkle tree database prefix in `lisk-chain`
- Update sparse merkle tree leaf value to only contain leaf prefix to avoid duplicate data - https://github.com/LiskHQ/lisk-sdk/issues/6900
- Update state store `finalize` to update sparse merkle tree
- Calculate state root at the end of block generation and add it to block header
- Add verification of the state root against block header in consensus
- Store SMT at the end of block execution
- Add re-calculation of sparse merkle tree for deleting block

### Acceptance Criteria
- Add unit tests
- Add integration test to show generation and execution produce the same stateRoot
- Add integration test to show deletion of the block will revert the state root to the previous block

### Additional Information
Related Issue #6749 

- sparse merkle tree should not be part of `diff`
- sparse merkle tree should use the newly defined database prefix

for reverting, diff can be used like below.
```
for (const kv of updated) {
  await smt.update(kv.key, kv.value);
}
for (const kv of removed) {
  await smt.remove(kv.key, kv.value);
}
```

### PR DESCRIPTION
**Pull request by ishantiw**
### What was the problem?

This PR resolves #6752 

### How was it solved?

- [x] Define sparse merkle tree database prefix in lisk-chain - https://github.com/LiskHQ/lisk-sdk/commit/c928e8b482ef7ae1d3a300d750e0e5e6d74daddc
- [x] Update state store finalize to update sparse merkle tree - https://github.com/LiskHQ/lisk-sdk/commit/1fe0910e953052a09e65d3658dc1693bc15ce1ce
- [x] Calculate state root at the end of block generation and add it to block header - https://github.com/LiskHQ/lisk-sdk/commit/b8e8a49f84f29f1e7e3182ff3833d5b25966cf37
- [x] Add verification of the state root against block header in consensus - https://github.com/LiskHQ/lisk-sdk/commit/c5be978dac56c32c7634c89d0d1072bed852b86f
- [x] Store SMT at the end of block execution - https://github.com/LiskHQ/lisk-sdk/commit/f1540da91307f6aff035f06b6d0d911faf8706ff
- [x] Add re-calculation of sparse merkle tree for deleting block - https://github.com/LiskHQ/lisk-sdk/commit/f1540da91307f6aff035f06b6d0d911faf8706ff
- [x] Add unit tests https://github.com/LiskHQ/lisk-sdk/pull/6899/commits/673326d9d302a892f75f5dcc01b56c545b29ab9f
- [x] Add integration test to show generation and execution produce the same stateRoot https://github.com/LiskHQ/lisk-sdk/pull/6899/commits/e786c32be70c8e899f01f727b20b3a17a2157db0
- [x] Add integration test to show deletion of the block will revert the state root to the previous block https://github.com/LiskHQ/lisk-sdk/pull/6899/commits/e786c32be70c8e899f01f727b20b3a17a2157db0

### How was it tested?

In framework,
`npm run test:unit generator`
`npm run test:unit consensus`
In lisk-chain,
`npm run test`

### CODE REVIEW

File: framework/src/node/consensus/consensus.ts [hunk 1]
```diff
@@ -641,17 +648,58 @@ export class Consensus {
...
 
 		// Offset must be set to 1, because lastBlock is still this deleting block
 		const stateStore = new StateStore(this._db);
-		await this._chain.removeBlock(block, stateStore, { saveTempBlock });
+		const currentState = await this._finalizeStore(stateStore);
+		await this._chain.removeBlock(block, currentState, { saveTempBlock });
 		this.events.emit(CONSENSUS_EVENT_BLOCK_DELETE, block);
 	}
 
+	private async _finalizeStore(stateStore: StateStore, finalize = true): Promise<CurrentState> {
```
Comment (shuse2) [inline framework/src/node/consensus/consensus.ts:671]: do we need `, finalize = true`? name is finalize and having flag seems bit strange.

File: framework/test/unit/modules/bft/bft_params.spec.ts [hunk 2]
```diff
@@ -65,7 +66,9 @@ describe('BFT parameters', () => {
 			};
 			await paramsStore.set(height2Bytes, codec.encode(bftParametersSchema, bftParams2));
 			const batch = db.batch();
-			rootStore.finalize(batch);
+			const smtStore = new SMTStore(db);
+			const smt = new SparseMerkleTree({ db: smtStore, keyLength: 11 });
```
Comment (shuse2) [inline framework/test/unit/modules/bft/bft_params.spec.ts:70]: `keyLength` is optional, so maybe we dont need it here to specify

File: elements/lisk-chain/src/data_access/data_access.ts [hunk 3]
```diff
@@ -19,8 +19,8 @@ import { Block } from '../block';
 
 import { BlockCache } from './cache';
 import { Storage as StorageAccess } from './storage';
-import { StateStore } from '../state_store';
 import { BlockAssets } from '../block_assets';
+import { CurrentState } from '..';
```
Comment (shuse2) [inline elements/lisk-chain/src/data_access/data_access.ts:23]: prefer to specify exact file

**Review by shuse2 — CHANGES_REQUESTED** [review]

File: framework/src/node/consensus/consensus.ts (hunk 1 above)
Comment (ishantiw) [reply framework/src/node/consensus/consensus.ts:671]: maybe `_prepareStore`, and then we can pass finalize flag ?

File: framework/test/unit/modules/bft/bft_params.spec.ts (hunk 2 above)
Comment (ishantiw) [reply framework/test/unit/modules/bft/bft_params.spec.ts:70]: In the tests, we are using different keyLengths and it throws error so I had to configure it

Comment (shuse2) [reply framework/test/unit/modules/bft/bft_params.spec.ts:70]: make sense!

File: elements/lisk-chain/src/data_access/storage.ts [hunk 4]
```diff
@@ -29,6 +28,7 @@ import {
 } from '../db_keys';
 import { concatDBKeys } from '../utils';
 import { stateDiffSchema } from '../schema';
+import { CurrentState } from '..';
```
Comment (shuse2) [inline elements/lisk-chain/src/data_access/storage.ts:31]: I think there are several places using this

File: framework/src/node/consensus/consensus.ts [hunk 5]
```diff
@@ -451,7 +454,14 @@ export class Consensus {
 			finalizedHeight = bftVotes.maxHeightPrecommitted;
 		}
 
-		await this._chain.saveBlock(block, stateStore, finalizedHeight, {
+		// Result Validation
+		// Verify validatorsHash
+		await this._verifyValidatorsHash(apiContext, block);
+		// Verify stateRoot
+		const currentState = await this._prepareStore(stateStore);
```
Comment (shuse2) [inline framework/src/node/consensus/consensus.ts:461]: I cannot  think of nicer name, but `currentState`  and `_prepareStore` still doesnt match. maybe `_prepareFinallizingState` would be bit better?

**Review by shuse2 — APPROVED** [review]

File: framework/src/node/consensus/consensus.ts (hunk 5 above)
Comment (ishantiw) [reply framework/src/node/consensus/consensus.ts:461]: yeah I think `_prepareFinalizingState` store sounds better

File: framework/test/unit/node/consensus/consensus.spec.ts [hunk 6]
```diff
@@ -506,21 +514,32 @@ describe('consensus', () => {
...
+				jest.spyOn(stateMachine, 'verifyBlock').mockResolvedValue();
+				jest.spyOn(stateMachine, 'executeBlock').mockResolvedValue();
+				jest.spyOn(bftAPI, 'getBFTParameters').mockResolvedValue({
+					validatorsHash: block.header.validatorsHash,
+				} as never);
+				jest.spyOn(consensus.events, 'emit');
+				consensus['_db'] = dbMock;
+				verifyStateRootMock = jest.fn();
+				consensus['_verifyStateRoot'] = verifyStateRootMock;
+				verifyStateRootMock.mockReturnValue(undefined);
```
Comment (mehmetegemen) [inline framework/test/unit/node/consensus/consensus.spec.ts:534]: Is there a reason why you didn't use something like `jest.spyOn().mockImplementation()` here so we can reset the mock when we need in a test case?

File: elements/lisk-chain/test/integration/data_access/blocks.spec.ts [hunk 7]
```diff
@@ -483,4 +514,198 @@ describe('dataAccess.blocks', () => {
...
+
+		const prefixedKey = (
+			key: Buffer,
+			moduleID: number = MODULE_ID,
+			storePrefix: number = STORE_PREFIX,
+		) => {
+			const moduleIDBuffer = Buffer.alloc(4);
+			moduleIDBuffer.writeInt32BE(moduleID, 0);
+			const storePrefixBuffer = Buffer.alloc(2);
+			storePrefixBuffer.writeUInt16BE(storePrefix, 0);
```
Comment (mehmetegemen) [inline elements/lisk-chain/test/integration/data_access/blocks.spec.ts:545]: is `intToBuffer` applicable here?

File: elements/lisk-chain/test/integration/data_access/blocks.spec.ts [hunk 8]
```diff
@@ -483,4 +514,198 @@ describe('dataAccess.blocks', () => {
...
+			stateStore = new StateStore(db);
+			tempDB = new InMemoryKVStore();
+		});
+
+		afterAll(async () => {
+			await tempDB.clear();
+		});
+
+		it('should return current stateRoot calculated after saveBlock', async () => {
+			const subStore = stateStore.getStore(14, 1);
```
Comment (mehmetegemen) [inline elements/lisk-chain/test/integration/data_access/blocks.spec.ts:560]: Is it okay to not use `STORE_PREFIX` and `MODULE_ID` we defined just above?

**Review by mehmetegemen — APPROVED** [review]
Minor comments, LGTM 🎉

File: framework/test/unit/node/consensus/consensus.spec.ts (hunk 6 above)
Comment (ishantiw) [reply framework/test/unit/node/consensus/consensus.spec.ts:534]: jest.spyOn() doesn't work for private function

Comment (mehmetegemen) [reply framework/test/unit/node/consensus/consensus.spec.ts:534]: I think I made it work like `jest.spyOn(class, '_privateMethod' as any).mockImplementation()` but it's open to a debate whether it's a good practice 😄 Got it.

Comment (ishantiw) [reply framework/test/unit/node/consensus/consensus.spec.ts:534]: ah nice, I mean if it works then lets use it from next time

**Pull request merged**